### PR TITLE
[FLINK-20325][build] Move docs_404_check to CI stage

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Currently, `docs_404_check` is only triggered nightly. We should move the check into CI stage to forbidding broken docs commits. 

## Brief change log

- Just move the docs_404_check to the "ci" stage in "build-apache-repo.yml"

## Verifying this change

N/A

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wuchong** [comment]
I converted this into draft. Will convert it back until the skip docs_check is verified in my private Azure.

**Review by rmetzger — APPROVED** [review]
I haven't tested it on my personal CI. Once you've validated that this works as expected, +1 to merge.

Thanks a lot for taking care of this!

**Comment by wuchong** [comment]
Hi @rmetzger , the build should fail because there are some 404 pages, however it is passed. 
I looked into the log, it did report an unfounded page by the web server process

```
[2020-11-25 16:48:48] ERROR `/zh/dev/datastream_execution_mode.html' not found.
```

, but the `check_links.sh` returns success and no broken links are found. 

I'm wondering maybe it is related to the `grep` command or the language of the `spider.log`.

Do you know how to download the `spider.log`?


Logs: 
```
 Auto-regeneration: enabled for '/home/vsts/work/1/s/docs'
    Server address: http://0.0.0.0:4000/
  Server running... press ctrl-c to stop.
Waiting for server...
HTTP/1.1 200 OK 
Etag: 8f45f-8418-5fbe8ae3
Content-Type: text/html; charset=utf-8
Content-Length: 33816
Last-Modified: Wed, 25 Nov 2020 16:48:35 GMT
Cache-Control: private, max-age=0, proxy-revalidate, no-store, no-cache, must-revalidate
Server: WEBrick/1.3.1 (Ruby/2.4.10/2020-03-31)
Date: Wed, 25 Nov 2020 16:48:40 GMT
Connection: Keep-Alive

[2020-11-25 16:48:48] ERROR `/zh/dev/datastream_execution_mode.html' not found.
All links in docs are valid!
```

https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=10128&view=logs&j=c5d67f7d-375d-5407-4743-f9d0c4436a81&t=38411795-40c9-51fa-10b0-bd083cf9f5a5

**Comment by HuangXingBo** [comment]
Many Chinese docs newly added recently have links to English documents that cause `build_doc.sh` to fail. But the existing logic of `check_docs.sh` is to test whether there are broken links, which cannot cover the problem of linking Chinese documents to English documents. I directly executed the script `./tools/ci/docs.sh` locally to test, and the generated `spider.log` will be in the flink directory. @wuchong , our existing Azure Pipeline does not upload `spider.log`. If you want to download it, you can add a task `PublishPipelineArtifact@1` to upload `spider.log`, and then you can download it.

**Comment by wuchong** [comment]
Hi @HuangXingBo ,  `./tools/ci/docs.sh` will run `build_doc.sh` first, so I think this can cover the problem you mentioned of linking Chinese documents to English documents? 

Thanks for the suggestions, I will try to upload the artifact first.

**Comment by HuangXingBo** [comment]
Hi @wuchong , In the script `docs.sh`, the doc is constructed by executing `./docs/build_docs.sh -p`, but the `-p` parameter will not build the doc content by language. We can execute `build_doc.sh` without parameters to build language-specific doc content, so as to check whether there is a cross-language link problem.
Maybe we can add a pre-step of executing the scrpt `build-doc.sh` parameterless build in the `docs.sh` script, and then executing original step of `build-doc.sh -p` to start the web server after success. Or we can directly modify the content of the parameters received by `build_doc.sh`

**Comment by rmetzger** [comment]
@dianfu can you take a look as well? Afaik you've recently fixed the link check script.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -75,6 +75,21 @@ stages:
...
+            inputs:
+              versionSpec: '= 2.4'
+              addToPath: true
+          - script: ./tools/ci/docs.sh
+          # upload spider.log for debugging
+          - task: PublishPipelineArtifact@1
+            inputs:
+              targetPath: $(FLINK_ARTIFACT_DIR)
+              artifact: spider.log
+              condition: not(eq(variables['SKIP'], '1'))
```
Comment (dianfu) [inline azure-pipelines.yml:92]: it seems that this line is unnecessary?

File: tools/azure-pipelines/build-apache-repo.yml [hunk 2]
```diff
@@ -68,6 +68,35 @@ stages:
...
+              versionSpec: '= 2.4'
+              addToPath: true
+          - script: ./tools/ci/docs.sh
+            condition: not(eq(variables['SKIP'], '1'))
+          # upload spider.log for debugging
+          - task: PublishPipelineArtifact@1
+            inputs:
+             targetPath: $(FLINK_ARTIFACT_DIR)
+             artifact: spider.log
+             condition: not(eq(variables['SKIP'], '1'))
```
Comment (dianfu) [inline tools/azure-pipelines/build-apache-repo.yml:99]:
```suggestion
           condition: not(eq(variables['SKIP'], '1'))
```

**Review by dianfu — COMMENTED** [review]
I have run the check_links.sh locally and found the following log in the spider.log:
{code}
Found 1 broken link.
{code}

So I guess it's because there is just one broken link and so "Found xx broken links" doesn't match it. We could just remove the "s" suffix.

**Review by dianfu — APPROVED** [review]

**Comment by rmetzger** [comment]
I believe we should try to get this in soon, the docs are constantly broken (I broke them too)

**Comment by rmetzger** [comment]
I'm closing this PR in favor of #14391

**Pull request closed without merging**